[FLINK-10156][table] Deprecate Table.writeToSink()

fhueske opened the pull request:
## What is the purpose of the change

* Deprecate `Table.writeToSink()` as discussed in FLINK-10156

## Brief change log

* Deprecate `Table.writeToSink()`
* Port usage of `writeToSink()` to `insertInto()`

## Verifying this change

* run existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? Docs have been adjusted.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala:
```diff
@@ -891,7 +891,13 @@ class Table(
     * @tparam T The data type that the [[TableSink]] expects.
+    *
+    * @deprecated Will be removed in a future release. Please register the TableSink and use
+    *             [[insertInto()]].
```
For the complete javadoc, we need to remove "[[", 
[[insertInto()]] -> insertInto()
![image](https://user-images.githubusercontent.com/22488084/46772885-efc94400-cd2c-11e8-81e2-9c488fa20785.png)
![image](https://user-images.githubusercontent.com/22488084/46772899-01aae700-cd2d-11e8-96f3-842f65058d31.png)

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala:
```diff
@@ -944,7 +956,12 @@ class Table(
+      case _: LogicalTableFunctionCall =>
+        throw new ValidationException("TableFunction can only be used in join and leftOuterJoin.")
+      case _ =>
+        tableEnv.insertInto(this, tableName, this.tableEnv.queryConfig)
```
Can we using "insertInto(tableName, this.tableEnv.queryConfig)", that share the validation logic ?

fhueske replied:
This doesn't work because `this.tableEnv` is null if the Table was generated from a `TableFunctionCall`. 
So we cannot call `insertInto(tableName, this.tableEnv.queryConfig)` because it throws an NPE.

sunjincheng121 replied:
I feel like that put same code in the two `overload` method is not the best way.  We should check the NULL of `tableEnv` Before we call `insertInto(tableName, this.tableEnv.queryConfig)`, It's that make sense to you?

fhueske replied:
The check is basically replacing the null check but provides more context. 
What would we do if `tableEnv` is `null`? Do nothing? Throw an exception? What  would be the error message? By checking for the type of the root node of the logical plan we know why the call would not succeed and can give an appropriate error message. 

We could of course also pass `null` as configuration and check for it to be null in `insertInto(String, QueryConfig)` but I'd consider that as an invalid use of the interface that should also result in an exception.

sunjincheng121 replied:
if we check NULL, we should throw exception, but I think your right. I agree keep the current approach .

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala:
```diff
@@ -960,7 +977,12 @@ class Table(
-    tableEnv.insertInto(this, tableName, conf)
+    this.logicalPlan match {
+      case _: LogicalTableFunctionCall =>
+        throw new ValidationException("TableFunction can only be used in join and leftOuterJoin.")
```
new ValidationException(..) ->  ValidationException(...)

sunjincheng121 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/AggregateITCase.scala:
```diff
@@ -307,11 +308,16 @@ class AggregateITCase extends StreamingWithStateTestBase {
     val t = env.fromCollection(data).toTable(tEnv, 'a, 'b, 'c)
             .groupBy('c)
-            .select('c, 'b.max)
+            .select('c, 'b.max as 'bMax)
```
Do we have to add alias ?

fhueske replied:
No, you are right. I assumed that field names need to match.

sunjincheng121 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/JoinITCase.scala:
```diff
@@ -26,6 +26,7 @@ import org.apache.flink.table.runtime.utils.{StreamITCase, StreamTestData, Strea
 import org.junit.Assert._
```
Please remove useless [[TableException]] import at line 23.

sunjincheng121 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/TableSinkITCase.scala:
```diff
@@ -177,10 +195,16 @@ class TableSinkITCase extends AbstractTestBase {
       .groupBy('len)
-      .select('len, 'id.count, 'num.sum)
-      .writeToSink(new TestRetractSink)
+      .select('len, 'id.count as 'icnt, 'num.sum as 'nsum)
```
In order to show that the column name of the source table can be inconsistent with the column name of the sink table, I recommend removing the alias. What do you think？

fhueske replied:
I'd keep it here for clarity.
There are other tests that check that field names do not need to match like `TableSinkITCase.testInsertIntoRegisteredTableSink()`.

sunjincheng121 replied:
Sounds good to me.

fhueske commented:
Thanks for the review @sunjincheng121!
I addressed your comments. 

I'd like to keep the `writeToSink` call in `LocalExecutor` for now and solve that in a follow up issue. We would need to register a temporary table, that cannot be removed anymore. There are plans to manage the catalog of the SQL Client into an `InMemoryExternalCatalog" that would make this easier.

sunjincheng121 commented:
Thanks for the update @fhueske !
The current change LGTM.
+1 to merged.

fhueske commented:
Thanks for the review @sunjincheng121!

fhueske commented:
merging
